zerver/tests/test_upload.py

```diff
@@ -395,6 +413,128 @@ def test_upload_size_quote(self):
+
+        # Confirm other cross-realm users can't read it.
+        self.login(user3_email, 'test')
+        response = self.client_get(uri)
+        self.assertEqual(response.status_code, 403)
+        self.assert_in_response("You are not authorized to view this file.", response)
+
+    def test_file_download_authorization_invite_only(self):
+        # type: () -> None
+        subscribed_users = ["hamlet@zulip.com", "iago@zulip.com"]
+        unsubscribed_users = ["othello@zulip.com", "prospero@zulip.com"]
+        for user in subscribed_users:
+            self.subscribe_to_stream(user, "test-subscribe")
+
+        # Make the stream private
+        stream = Stream.objects.get(name='test-subscribe')
+        stream.invite_only = True
+        stream.save()
+
+        self.login("hamlet@zulip.com")
+        fp = StringIO("zulip!")
+        fp.name = "zulip.txt"
+        result = self.client_post("/json/upload_file", {'file': fp})
+        json = ujson.loads(result.content)
+        uri = json["uri"]
+        fp_path_id = re.sub('/user_uploads/', '', uri)
+        body = "First message ...[zulip.txt](http://localhost:9991/user_uploads/" + fp_path_id + ")"
+        self.send_message("hamlet@zulip.com", "test-subscribe", Recipient.STREAM, body, "test")
+        self.client_post('/accounts/logout/')
+
+        # Subscribed user should be able to view file
+        for user in subscribed_users:
+            self.login(user)
+            response = self.client_get(uri)
+            self.assertEqual(response.status_code, 200)
+            data = b"".join(response.streaming_content)
+            self.assertEqual(b"zulip!", data)
+            self.client_post('/accounts/logout/')
+
+        # Unsubscribed user should not be able to view file
+        for user in unsubscribed_users:
+            self.login(user)
+            response = self.client_get(uri)
+            self.assertEqual(response.status_code, 403)
+            self.assert_in_response("You are not authorized to view this file.", response)
+            self.client_post('/accounts/logout/')
+
+    def test_file_download_authorization_public(self):
+        # type: () -> None
+        subscribed_users = ["hamlet@zulip.com", "iago@zulip.com"]
+        unsubscribed_users = ["othello@zulip.com", "prospero@zulip.com"]
+        for user in subscribed_users:
+            self.subscribe_to_stream(user, "test-subscribe")
+
+        self.login("hamlet@zulip.com")
+        fp = StringIO("zulip!")
+        fp.name = "zulip.txt"
+        result = self.client_post("/json/upload_file", {'file': fp})
+        json = ujson.loads(result.content)
+        uri = json["uri"]
+        fp_path_id = re.sub('/user_uploads/', '', uri)
+        body = "First message ...[zulip.txt](http://localhost:9991/user_uploads/" + fp_path_id + ")"
+        self.send_message("hamlet@zulip.com", "test-subscribe", Recipient.STREAM, body, "test")
+        self.client_post('/accounts/logout/')
```

Comment:
I said `self.logout()`, but what Django actually supports is`self.client.logout()`:

https://docs.djangoproject.com/en/1.11/topics/testing/tools/#django.test.Client.logout

I think what we should do here is parallel what we did with `login`.  We have a `login` wrapper in ZulipTestCase that essentially calls `self.client.login` (with a few other minor details).  We should have a similar wrapper for `logout`.  This way tests that don't really care about the details of our logging-out mechanism don't need to specify the exact URL we use.